Simplify value modification (#2747)
modelica/ModelicaSpecification: RationaleMCP/0031/differences.md

# Semantical differences between Flat Modelica and Modelica
This document describes differences between Flat Modelica and Modelica that aren't clear from the differences in the grammars.

## Top level structure

The top level structure (see [grammar](grammar.md#Start-rule)) of a Flat Modelica description can have several top level definitions, with a mandatory `model` definition at the end.
The definitions before the `model` either define types or global constants.

## Lexical scoping and record definitions

Lookup in Flat Modelica is significantly simplified compared to full Modelica due to the restricted top level structure of a Flat Modelica program, but there are two more restrictions on top of that explained in this section.

Taken together, the two restrictions can be summarized concisely as follows:

> - In Flat Modelica, members of a record definition are never in lexical scope.
>   In other words, record members can only be accessed through instances of the record.

Reviewer: I would propose to just have one line:
```suggestion
- In Flat Modelica, a member of a record can only be accessed through an instance of the record.
```
Having two restrictions that are summarized in two different ways was somewhat confusing. For lexical scoping there are different variants of restricting it (some languages allow lookup but make it illegal to use such cases), but it seems we implicitly have defined that all lookup start at the top-level package.

Author: Just for clarification, by "we", do you mean the Flat Modelica MCP group, or the Modelica specification?  I couldn't understand it as Modelica specification as it would only be for fully qualified component references (such as `.Modelica.Constants.pi`) that we start at the top (one level above the top level package, actually).  That leaves me with the Flat Modelica MCP group interpretation, and if this is right it seems like something worth noting explicitly where we talk about lookup.

However, doesn't the following example show that lookup still starts at the nearest enclosing scope, just like in full Modelica?
```
package 'MyModel'
  constant Real 'x' = 0.1;
  model 'MyModel'
    Real 'x' = 0.2;
    Real 'y' = 'x' + 1.0; /* This is the 'x' in the model, not the global constant. */
  end 'MyModel';
end 'MyModel';
```

Reviewer 2: Web Meeting: Hans, Henrik, Martin, Oliver
```suggestion
- In Flat Modelica, a member of a record can only be accessed through an instance of the record.
  (This can also be described in terms of lexical look-up rules.)
```

# Semantical differences between Flat Modelica and Modelica
This document describes differences between Flat Modelica and Modelica that aren't clear from the differences in the grammars.

> ## Lexical scoping inside record definitions

Reviewer 2: Should contain two subsection covering
inside records
not treating records as packages

Author: Done.

```

The following restriction applies to modifications in types and functions, making types and function signatures in Flat Modelica easier to represent and reason about compared to full Modelica:

> - Attribute modifiers must have constant variability.

Reviewer 2: Add explicitly that value modifications in model component declarations are not affected.

Author: This is already stated in the proper place; this list is only about types and functions.

Reviewer 2: Resolved in web meeting
